klifs_utils/local/coordinates/water.py

```diff
@@ -0,0 +1,44 @@
+"""
```

Comment:
All these modules under `coordinates/` are very similar. Although you have extracted the code to utility functions (well done!), I think we can come up with something even more maintainable. I'll think about this and propose something next. Not important for now. Consider this more of a personal note to myself :)

Reply:
I could have a module `klifs_utils/remote/util.py` with the following functions:

```python
def _coordinates_mol_text(structure_id, structure_type):
    """
    Get structural data content (mol file) from KLIFS database as string (text).
    
    Parameters
    ---------------
    structure_id : str
        KLIFS structure ID.
    structure_type : str
        Type of structural data, i.e. complex, protein, pocket, or ligand.

    Returns
    ----------
    str
        Structural data.
    """
    pass
```

That could remote all those functions:
```
klifs_utils.remote.coordinates.complex._complex_mol2_text
```

Reply:
I am adding here all `coordinate` functions from the `remote` and `local` module, so that I (and maybe you too) can think more easily about other options to reduce redundancy. 

#### `remote` module

Parameter for these functions is:
`structure_id`

```
klifs_utils.remote.coordinates.complex.mol2_to_dataframe
klifs_utils.remote.coordinates.complex.pdb_to_dataframe
klifs_utils.remote.coordinates.ligand.mol2_to_dataframe
klifs_utils.remote.coordinates.ligand.mol2_to_rdkit_mol
klifs_utils.remote.coordinates.pocket.mol2_to_dataframe
klifs_utils.remote.coordinates.protein.mol2_to_dataframe
```

#### `local` module

Parameter for these functions are:
`klifs_download_path`
`species`
`kinase_name`
`pdb_id`
`alt`
`chain`

Note: The KLIFS manual download does not provide KLIFS IDs such as the `structure_id`. However, I was thinking that I could add a function to the `local` module that does fetch these IDs from KLIFS, so that the functions below could have the same parameter structure as in the `remote` module.

```
klifs_utils.local.coordinates.complex.mol2_to_dataframe
klifs_utils.local.coordinates.ligand.mol2_to_dataframe
klifs_utils.local.coordinates.ligand.mol2_to_rdkit_mol
klifs_utils.local.coordinates.pocket.mol2_to_dataframe
klifs_utils.local.coordinates.protein.mol2_to_dataframe
```

Reply:
> How to load klifs_metadata in local module more elegantly (can this be set as global variable)?
> Can you elaborate a bit more? What's the current approach?

Yes, that thought was not described properly, sorry. 

When downloading KLIFS data manually, we get two csv files with metadata information -  basically details per KLIFS structure entry (one entry = structure from PDB ID X, chain X and alternate model X). 

Using the [`klifs_utils.local.initialize`](https://github.com/volkamerlab/klifs_utils/blob/lib-skeleton/klifs_utils/local/initialize.py) function will merge these two files into a unified DataFrame (let's call it `klifs_metdata`) containing per KLIFS structure entry all metadata available in your local csv files.

Why bother? In the `kissim` project, I am using `klifs_metadata` to filter the dataset for e.g. human DFG-in structure entries (`klifs_metadata_filtered`). Then, I am using `klifs_metadata_filtered` to iterate over all my selected structures - it is my link to the local mol2 files plus additional information about kinase name/family/group, resolution, ect.

Back to `klifs_utils`: I was considering to use `klifs_metadata` in the following modules to query kinases/ligands/interactions metadata in the local KLIFS data
```
klifs_utils.local.interactions
klifs_utils.local.kinases
klifs_utils.local.ligands
klifs_utils.local.structures
```
similar to what we do it in
```
klifs_utils.remote.interactions
klifs_utils.remote.kinases
klifs_utils.remote.ligands
klifs_utils.remote.structures
```
with questions/actions like
* What kind of kinases do I have in my local dataset? (KLIFS allows you to filter data before download, so this question will result in different answers depending on the kind of download the user did)
* Show me the interaction fingerprint or kinase pocket sequence for certain structures.
* Show me all KLIFS entries in my dataset for ABL1.

Back to your initial question:
For all functions that deal with these questions, I currently will need to pass `klifs_metadata` as input parameter. But I was hoping that I can (after running `klifs_utils.local.initialize` and maybe doing some filtering) load `klifs_metadata` somewhere globally?
Example functions exist here: 
https://github.com/volkamerlab/klifs_utils/blob/lib-skeleton/klifs_utils/local/kinases.py

Reply:
I see. So you need some kind of _global_ state. I'd look into something similar to what `requests` does.

They have global functions that behave with default parameters, like `requests.get` or `requests.post`. However, if you want to change the options, you can instantiate a session like this:

```
session = requests.Session(<custom options>)
#  and use session.<method> instead of requests.<method>
session.get(url)
```

Behind the scenes, `requests` creates a default `Session()` at the first import time, and reuses that later. You could do something similar? Let me know if you need more pointers!

Reply:
This would require a slight OOP refactor, btw 😬